// @textlint/kernel/lib/utils/src/KeyPath/TextlintKeyPath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TextlintPackagePrefix_1 = require("./TextlintPackagePrefix");
var KeyPathUtil_1 = require("./KeyPathUtil");
/**
 * normalize `keyPath` that is specific path for rule
 * This normalize function handle ambiguity `key`
 * `keyPath` is one of "preset/rule` key, or "rule" key
 *
 * Note: `textlint-rule-textlint-rule-x`(twice!) is normalized at once.
 * It is edge case and this function can not treat it.
 * @param keyPath
 */
exports.normalizeTextlintKeyPath = function (keyPath) {
    var _a = KeyPathUtil_1.splitKeyToPresetSubRule(keyPath), preset = _a.preset, rule = _a.rule;
    if (!preset) {
        return exports.normalizeTextlintRuleKey(rule);
    }
    return exports.normalizeTextlintRulePresetKey(preset) + "/" + exports.normalizeTextlintRuleKey(rule);
};
/**
 * Normalize to preset-name/rule-name
 */
exports.normalizeTextlintPresetSubRuleKey = function (names) {
    var preset = names.preset, rule = names.rule;
    return exports.normalizeTextlintRulePresetKey(preset) + "/" + exports.normalizeTextlintRuleKey(rule);
};
/**
 * Normalize rule key
 * @param name
 */
exports.normalizeTextlintRuleKey = function (name) {
    return KeyPathUtil_1.removePrefixFromPackageName([TextlintPackagePrefix_1.TextlintPackageNamePrefix.rule], name);
};
/**
 * Normalize filter rule key
 * @param name
 */
exports.normalizeTextlintFilterRuleKey = function (name) {
    return KeyPathUtil_1.removePrefixFromPackageName([TextlintPackagePrefix_1.TextlintPackageNamePrefix.filterRule], name);
};
/**
 * Normalize rule preset key
 * @param name
 */
exports.normalizeTextlintRulePresetKey = function (name) {
    // "preset-<name>" and "textlint-rule-preset-"
    return KeyPathUtil_1.removePrefixFromPackageName([TextlintPackagePrefix_1.TextlintPackageNamePrefix.rulePreset, "preset-"], name);
};
/**
 * Normalize plugin key
 * @param name
 */
exports.normalizeTextlintPluginKey = function (name) {
    return KeyPathUtil_1.removePrefixFromPackageName([TextlintPackagePrefix_1.TextlintPackageNamePrefix.plugin], name);
};
