// commandpost/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = __importDefault(require("./command"));
exports.Command = command_1.default;
const option_1 = __importDefault(require("./option"));
exports.Option = option_1.default;
const argument_1 = __importDefault(require("./argument"));
exports.Argument = argument_1.default;
var error_1 = require("./error");
exports.CommandpostError = error_1.CommandpostError;
exports.ErrorReason = error_1.ErrorReason;
/**
 * Create new top level command.
 * @param cmdName
 * @returns {Command<Opt, Arg>}
 */
function create(cmdName) {
    return new command_1.default(cmdName);
}
exports.create = create;
/**
 * exec parsing and call callback.
 * @param cmd it created by create function.
 * @param argv pass process.argv
 * @returns {Promise<{}>}
 */
function exec(cmd, argv) {
    return Promise
        .resolve(null)
        .then(() => {
        argv = argv.slice(2);
        // cmd.parse throw an exception often.
        return cmd.parse(argv);
    });
}
exports.exec = exec;
