// textlint-plugin-asciidoctor/lib/AsciidoctorProcessor.js
// LICENSE : MIT
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AsciidoctorProcessor = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _parse = require("./parse");

var _parse2 = _interopRequireDefault(_parse);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var AsciidoctorProcessor = exports.AsciidoctorProcessor = function () {
  function AsciidoctorProcessor(config) {
    _classCallCheck(this, AsciidoctorProcessor);

    this.config = config;
  }

  _createClass(AsciidoctorProcessor, [{
    key: "processor",
    value: function processor(ext) {
      return {
        preProcess: function preProcess(text, filePath) {
          return (0, _parse2.default)(text);
        },
        postProcess: function postProcess(messages, filePath) {
          return {
            messages: messages,
            filePath: filePath ? filePath : "<asciidoc>"
          };
        }
      };
    }
  }], [{
    key: "availableExtensions",
    value: function availableExtensions() {
      return [".adoc", ".asciidoctor", ".asc"];
    }
  }]);

  return AsciidoctorProcessor;
}();
